### ISSUE TITLE
[7.x] Incorrect sequence of quest objectives in packet SMSG_QUERY_QUEST_INFO_RESPONSE and SMSG_QUEST_GIVER_QUEST_DETAILS

### ISSUE THREAD

**Issue by Druson** [TrinityCore/TrinityCore#18061]
[//]: # (***************************************************************)
[//]: # (** DON'T DELETE THIS TEMPLATE ELSE YOUR ISSUE WILL BE CLOSED **)
[//]: # (***************************************************************)

**Description**:
The quest objectives can not be shown in correct sequence.

**Current behaviour**: Tell us what happens
For the quests added in 7.x, after quest objective completed, the next quest objective  will be shown in the tracing panel and quest detail dialog automatically. But the quest objectives can not be shown in correct sequence, and some quest objectives can not be traced in the tracing panel and the POI can not be shown in the map as well.

**Expected behaviour**: Tell us what should happen instead
The quest objectives can be shown in correct sequence in the tracing panel or the quest detail dialog.

**Steps to reproduce the problem**:
Use quest http://www.wowhead.com/quest=40378
1. .quest add 40378
2. The first quest objective [Ashtongue forces] can not be shown, but the third objective [Accept Illidan's gift] was shown in the tracing panel.
3. Checked the sniffing data (from the blz server) in packet SMSG_QUERY_QUEST_INFO_RESPONSE and SMSG_QUEST_GIVER_QUEST_DETAILS, the sequence of the objectives in the packet are:
[0] - objective ID 280768, object ID 88872, StorageIndex 0 [Ashtongue forces]
[1] - objective ID 281158, object ID 101534, StorageIndex 4 [See Felsaber Hidden Objective]
[2] - objective ID 281156, object ID 101518, StorageIndex 3 [Accept Illidan's gift]
[3] - objective ID 281030, object ID 94410, StorageIndex 2  [Find Allari to the southeast]
[4] - objective ID 280769, object ID 97831, StorageIndex 1 [First Summoned Guardian]

Changed the code to make the sequence of the objectives in the packet same as above, it worked as expected: [0] was seen firstly, after [0] completed, [1] auto completed, [2] should be seen, and after [2] completed, [3] should be seen, and after [3]  completed, [4] auto completed as well, the quest should be completed.

**Branch(es)**:  6.x 

**TC hash/commit**:  d1ae3ccb220f7582690dc849a8fb7bd37f4a2f18

**TDB version**:  6.04

**Operating system**:  Win10

[//]: # (This template is for problem reports, for other type of reports edit it accordingly)
[//]: # (If this is a crash report, include the crashlog with https://gist.github.com/)
[//]: # (For fixes containing c++ create a Pull Request)

**Comment by Druson** [comment]
I have made a fix by adding a new field "PacketSeqNo" in table quest_objectives, and update the code in
<https://github.com/TrinityCore/TrinityCore/blob/6.x/src/server/game/Globals/ObjectMgr.cpp#L3898> as below:
```
result = WorldDatabase.Query("SELECT ID, QuestID, Type, StorageIndex, ObjectID, Amount, Flags, ProgressBarWeight, Description FROM quest_objectives ORDER BY PacketSeqNo ASC, StorageIndex DESC");
```
And I have tested it in game and the result was as expected.

**Comment by Shauren** [comment]
This looks like the quest_poi issues we had a few years ago where packet needed to be sent with exactly the same data/order (which was also solved by adding a new column to that table)

**Comment by cocolino** [comment]
This problem from my research should not require any new db field. It is related to missing handling of quest objective flags 0x1 -> 0x40. For example, flag 0x2 is overlapped, meaning some objectives are hidden until the ones without this flag are complete, and displayed afterwards. The client handles the most part here but there is also another issue: the quest objectives are not loaded properly on server startup, they should be loaded in storageindex  ascending order (check "Load `quest_objectives` order by descending storage index to reduce resizes" in objectMgr -> that is wrong and causes the actual order display problem).

**Comment by Druson** [comment]
@cocolino The behaviour is not exactaly as what you checked actually. I have tested in game. If only load the objectives in ASC order, and put them in the packet in same order, two questobjectives can be seen at the same time.  For example , after questobjective [0] completed, you can find that both [2] and [3] can be seen at the same time, but the actually [3] should be seen only after [2] completed.

**Comment by cocolino** [comment]
@Druson That is because you need the objective flags properly + implement handling for them. 
This is no longer a loading issue, the loading is fine now.

Storage index 0 - Ashtongue forces - Flags 0.
Storage index 1 - First Summoned Guardian - Flags 28.
Storage index 2 - Find Allari - Flags 2.
Storage index 3 - Accept Illidan's Gift - Flags 0.
Storage index 4 - See Felsaber hidden obj - Flags 26.

First check if you haven't changed these in your own db please, these are default and should not be modified.

**Comment by Druson** [comment]
@cocolino Yes, I have set the objective flag correctly in db, actually the flag can be found in the sniffing data. And I have tried many objective orders in the packet aswell, but only the sequence which is from sniffing can act as expected. If client only determine the objectives order by the objective flags, any packet sequence should work as expected. So seems that the quest objective sequence in the packet has impact to the client behaviour.
